### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by miguelbiron** [Julia-Tempering/tennis_fun#1]
Hi @wongl01 -- I was looking at your code for the GPU tests and there are several things you could try to get better performance 

1. You seem to be confusing the `global` annotation with the `const` annotation (example [here](https://github.com/Julia-Tempering/tennis_fun/blob/a31c0566f545ca8c190341ad83a5bd7b5ef2a2cd/tennis_gpu.jl#L20)). The latter is the one you probably want to use. It tells the compiler that the type of a variable defined in Main is fixed, which speeds up things. The `global` annotation only is used to modify global vars in code written directly in the Main module outside of a function -- something you should never do.
2. You seem to be converting back and forth between Float64 and Float32. [This instruction](https://github.com/Julia-Tempering/tennis_fun/blob/a31c0566f545ca8c190341ad83a5bd7b5ef2a2cd/tennis_gpu.jl#L35) for example, where the broadcast even generates an undesirable allocation.
3. There are at least three allocations happening inside the logpotential function: thenone mentioned above, plus the ones from [this](https://github.com/Julia-Tempering/tennis_fun/blob/a31c0566f545ca8c190341ad83a5bd7b5ef2a2cd/tennis_gpu.jl#L37) and [this](https://github.com/Julia-Tempering/tennis_fun/blob/a31c0566f545ca8c190341ad83a5bd7b5ef2a2cd/tennis_gpu.jl#L39) temp vectors. You cpuld probably get away with doing all this work in place using a single for loop or mapreduce construct (if functional programming os needed for the GPU).

Let me know if you'd like to try and fix these issues and I can give you more detailed help.

**Comment by wongl01** [comment]
Hi Miguel, thanks for the advice, and sorry for taking so long to respond. I've pushed a new version that tried to address some of the issues. 
1. Swapped to using const instead of global now
2. I've made pigeons initialize a Float32 vector now, and randn! appears to keep the type, so this should be fixed too.
3. I've also attempted to address this by pre-allocating memory for the log likelihood and log prior vectors, then writing to them in the function, which should only come at a constant overhead but maybe doing it in place with mapreduce would still be faster.
